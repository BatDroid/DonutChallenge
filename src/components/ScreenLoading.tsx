import React from 'react';
import {ActivityIndicator} from 'react-native';

export const ScreenLoading = () => (
  <ActivityIndicator size="large" style={{alignSelf: 'center', flex: 1}} />
);
