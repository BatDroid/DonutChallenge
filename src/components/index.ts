export * from './ScreenLoading';
export * from './ServerError';
