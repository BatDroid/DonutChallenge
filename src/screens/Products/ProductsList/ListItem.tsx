import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  Dimensions,
  TouchableHighlight,
} from 'react-native';
import {ProductType} from '../../../redux/products/actions/types';

const {height} = Dimensions.get('window');
export const ITEM_HEIGHT = height * 0.068;

interface Props {
  item: ProductType;
  onItemPressed: (product: ProductType) => void;
}

export default ({item, onItemPressed}: Props) => {
  const {display_name, status} = item;
  const onPress = () => onItemPressed(item);
  return (
    <TouchableHighlight onPress={onPress} underlayColor="#cccc">
      <View style={styles.itemRoot}>
        <View style={styles.infoContainer}>
          <Text style={styles.displayName}>{display_name}</Text>
          <Text style={styles.status}>{status}</Text>
        </View>
        {/* todo: add chevron Icon from react-native-vector-icons */}
        <Text style={styles.chevron}>></Text>
      </View>
    </TouchableHighlight>
  );
};

const styles = StyleSheet.create({
  itemRoot: {
    height: ITEM_HEIGHT,
    borderBottomWidth: 0.5,
    borderColor: '#cccc',
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  infoContainer: {
    justifyContent: 'space-evenly',
    marginLeft: '2%',
  },
  displayName: {
    fontSize: 20,
  },
  chevron: {
    fontSize: 20,
    marginRight: '2%',
  },
  status: {
    color: '#666',
  },
});
