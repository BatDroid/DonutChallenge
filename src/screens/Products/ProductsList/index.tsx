import React from 'react';
import {FlatListProps, FlatList} from 'react-native';
import ListItem, {ITEM_HEIGHT} from './ListItem';
import {ProductType} from 'src/redux/products/actions/types';

interface Props extends Omit<FlatListProps<ProductType>, 'renderItem'> {
  onItemPressed: (prodcut: ProductType) => void;
}

export default (props: Props) => {
  const {onItemPressed, ...listProps} = props;
  return (
    <FlatList
      keyExtractor={i => i.id.toString()}
      initialNumToRender={18}
      renderItem={({item}) => {
        return <ListItem item={item} onItemPressed={onItemPressed} />;
      }}
      getItemLayout={(_, index) => ({
        length: ITEM_HEIGHT,
        offset: ITEM_HEIGHT * index,
        index,
      })}
      {...listProps}
    />
  );
};
