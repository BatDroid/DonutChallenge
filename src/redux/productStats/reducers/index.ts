import {combineReducers} from 'redux';
import {
  StatsType,
  GET_STATS_SUCCESS,
  GET_STATS_REQUEST,
  GET_STATS_ERROR,
  StatAction,
} from '../actions/types';

function statReducer(state: StatsType | null = null, action: StatAction) {
  switch (action.type) {
    case GET_STATS_SUCCESS:
      return action.stats;
    default:
      return state;
  }
}

function isFetchingReducer(state: boolean = false, action: StatAction) {
  switch (action.type) {
    case GET_STATS_REQUEST:
      return true;
    case GET_STATS_SUCCESS:
    case GET_STATS_ERROR:
      return false;
    default:
      return state;
  }
}

function errorReducer(state: number | null = null, action: StatAction) {
  switch (action.type) {
    case GET_STATS_REQUEST:
    case GET_STATS_SUCCESS:
      return null;
    case GET_STATS_ERROR:
      return action.errorCode;
    default:
      return state;
  }
}

export interface StatStoreType {
  errorCode: number | null;
  isFetching: boolean;
  stats: StatsType | null;
}

export default combineReducers({
  stats: statReducer,
  isFetching: isFetchingReducer,
  errorCode: errorReducer,
});
