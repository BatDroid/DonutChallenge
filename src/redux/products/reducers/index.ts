import {combineReducers} from 'redux';
import {
  ProductAction,
  ProductType,
  GET_PRODUCTS_SUCCESS,
  GET_PRODUCTS_REQUEST,
  GET_PRODUCTS_ERROR,
} from '../actions/types';

function productsReducer(
  state: ProductType[] | null = null,
  action: ProductAction,
) {
  switch (action.type) {
    case GET_PRODUCTS_SUCCESS:
      return action.products;
    default:
      return state;
  }
}

function isFetchingReducer(state: boolean = false, action: ProductAction) {
  switch (action.type) {
    case GET_PRODUCTS_REQUEST:
      return true;
    case GET_PRODUCTS_SUCCESS:
    case GET_PRODUCTS_ERROR:
      return false;
    default:
      return state;
  }
}

function errorReducer(state: number | null = null, action: ProductAction) {
  switch (action.type) {
    case GET_PRODUCTS_REQUEST:
    case GET_PRODUCTS_SUCCESS:
      return null;
    case GET_PRODUCTS_ERROR:
      return action.errorCode;
    default:
      return state;
  }
}

export interface ProductsStoreType {
  errorCode: number | null;
  isFetching: boolean;
  products: ProductType[] | null;
}

export default combineReducers({
  products: productsReducer,
  isFetching: isFetchingReducer,
  errorCode: errorReducer,
});
