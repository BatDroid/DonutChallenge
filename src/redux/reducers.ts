import {combineReducers} from 'redux';
import products from './products/reducers';
import productStats from './productStats/reducers';

export default combineReducers({
  products,
  productStats,
});
