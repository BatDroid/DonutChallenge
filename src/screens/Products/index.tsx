import React, {useEffect} from 'react';
import {SafeAreaView} from 'react-native';
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import {getProducts as getProuctsApi} from '../../redux/products/actions';
import {ProductsStoreType} from 'src/redux/products/reducers';
import {ServerError, ScreenLoading} from '../../components';
import ProductsList from './ProductsList';

interface Props extends ProductsStoreType {
  getProducts: Function;
}

const Products = (props: Props) => {
  const {getProducts, errorCode, isFetching, products} = props;
  useEffect(() => {
    getProducts();
  }, []);

  if (errorCode) return <ServerError errorMessage="unknown error" />;
  if (isFetching || !products) return <ScreenLoading />;

  return (
    <SafeAreaView>
      <ProductsList
        data={products}
        onItemPressed={product => {
          console.log(product);
        }}
      />
    </SafeAreaView>
  );
};

function mapDispatchToProps(dispatch: Dispatch) {
  return bindActionCreators(
    {
      getProducts: getProuctsApi,
    },
    dispatch,
  );
}

function mapStateToProps({
  products: {products, isFetching, errorCode},
}: {
  products: ProductsStoreType;
}) {
  return {
    products,
    isFetching,
    errorCode,
  };
}

const Connected: any = connect(
  mapStateToProps,
  mapDispatchToProps,
)(Products);

Connected.navigationOptions = {
  title: 'Products',
};

export default Connected;
