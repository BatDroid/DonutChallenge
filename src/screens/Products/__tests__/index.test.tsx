import 'react-native';
import React from 'react';
import thunk from 'redux-thunk';
import {shallow} from 'enzyme';
import configureStore from 'redux-mock-store';
import toJson from 'enzyme-to-json';
import {act} from 'react-test-renderer';
import {waitForElement} from '@testing-library/react-native';
import {render} from 'react-native-testing-library';
import Products from '../index';
import {ProductsStoreType} from '../../../redux/products/reducers';
import {ScreenLoading, ServerError} from '../../../components';
import ProductsList from '../ProductsList';
import store from '../../../configs/redux/store';

const middlewares = [thunk];

const mockStore = configureStore(middlewares);

const mockProducts = [
  {
    id: 'EOS-BTC',
    base_currency: 'EOS',
    quote_currency: 'BTC',
    base_min_size: '0.10000000',
    base_max_size: '5000.00000000',
    quote_increment: '0.00000100',
    base_increment: '0.10000000',
    display_name: 'EOS/BTC',
    min_market_funds: '0.001',
    max_market_funds: '30',
    margin_enabled: false,
    post_only: false,
    limit_only: true,
    cancel_only: false,
    status: 'online',
    status_message: '',
  },
];

describe('Products Screen', () => {
  it('renders correctly when component is Fetching Data', () => {
    const mockState: {
      products: ProductsStoreType;
    } = {
      products: {products: null, isFetching: true, errorCode: null},
    };

    const Component = shallow(<Products store={mockStore(mockState)} />)
      .dive()
      .dive();

    expect(Component.find(ScreenLoading)).toHaveLength(1);
    expect(toJson(Component.dive())).toMatchSnapshot();
  });

  it('renders correctly when component is throwing Error', () => {
    const mockState: {
      products: ProductsStoreType;
    } = {
      products: {products: null, isFetching: false, errorCode: 500},
    };

    const Component = shallow(<Products store={mockStore(mockState)} />)
      .dive()
      .dive();

    expect(Component.find(ServerError)).toHaveLength(1);
    expect(toJson(Component.dive())).toMatchSnapshot();
  });

  it('renders correctly when component recieves the data', () => {
    const mockState: {
      products: ProductsStoreType;
    } = {
      products: {products: mockProducts, isFetching: false, errorCode: null},
    };

    const Component = shallow(<Products store={mockStore(mockState)} />)
      .dive()
      .dive();

    expect(Component.find(ProductsList)).toHaveLength(1);
    expect(toJson(Component)).toMatchSnapshot();
  });

  it('renders properly when is Fetching data and showing data when data is recieved', async () => {
    const {getAllByType} = render(<Products store={store} />);
    await act(async () => {
      await waitForElement(() => getAllByType(ScreenLoading));
      expect(getAllByType(ScreenLoading)).toBeTruthy();
      await waitForElement(() => getAllByType(ProductsList));
      expect(getAllByType(ProductsList)).toBeTruthy();
    });
  });
});
