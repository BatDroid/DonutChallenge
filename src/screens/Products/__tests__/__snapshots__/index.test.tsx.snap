// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Screen renders correctly when component is Fetching Data 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "alignSelf": "center",
      "flex": 1,
    }
  }
/>
`;

exports[`Products Screen renders correctly when component is throwing Error 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "red",
      }
    }
  >
    unknown error
  </Text>
</View>
`;

exports[`Products Screen renders correctly when component recieves the data 1`] = `
<SafeAreaView>
  <_default
    data={
      Array [
        Object {
          "base_currency": "EOS",
          "base_increment": "0.10000000",
          "base_max_size": "5000.00000000",
          "base_min_size": "0.10000000",
          "cancel_only": false,
          "display_name": "EOS/BTC",
          "id": "EOS-BTC",
          "limit_only": true,
          "margin_enabled": false,
          "max_market_funds": "30",
          "min_market_funds": "0.001",
          "post_only": false,
          "quote_currency": "BTC",
          "quote_increment": "0.00000100",
          "status": "online",
          "status_message": "",
        },
      ]
    }
    onItemPressed={[Function]}
  />
</SafeAreaView>
`;
